web/src/main/java/org/cbioportal/web/StudyViewController.java

```diff
@@ -320,32 +320,69 @@ private void init() {
 
         clinicalAttributeUtil.extractCategorizedClinicalAttributes(clinicalAttributes, sampleAttributeIds, patientAttributeIds, patientAttributeIds);
 
+        List<Sample> samples = new ArrayList<>();
         List<Patient> patients = new ArrayList<>();
         List<String> patientIds = new ArrayList<>();
         List<String> studyIdsOfPatients = new ArrayList<>();
+        Map<String, Map<String, List<Sample>>> patientToSamples = null;
+        
 
         if (CollectionUtils.isNotEmpty(patientAttributeIds)) {
+            samples = sampleService.fetchSamples(studyIds, sampleIds, Projection.DETAILED.name());
             patients = patientService.getPatientsOfSamples(studyIds, sampleIds).stream().collect(Collectors.toList());
             patientIds = patients.stream().map(Patient::getStableId).collect(Collectors.toList());
             studyIdsOfPatients = patients.stream().map(Patient::getCancerStudyIdentifier).collect(Collectors.toList());
+            patientToSamples = samples.stream().collect(
+                Collectors.groupingBy(Sample::getPatientStableId, Collectors.groupingBy(Sample::getCancerStudyIdentifier))
+            );
         }
 
         List<ClinicalData> clinicalDataList = clinicalDataFetcher.fetchClinicalData(
             studyIds, sampleIds, patientIds, studyIdsOfPatients, sampleAttributeIds, patientAttributeIds, null
         );
-
-        Map<String, Map<String, List<ClinicalData>>> clinicalDataMap;
-        if (!sampleAttributeIds.isEmpty()) {
-            clinicalDataMap = clinicalDataList.stream().collect(Collectors.groupingBy(ClinicalData::getSampleId, 
-                Collectors.groupingBy(ClinicalData::getStudyId)));
+        
+        List<ClinicalData> sampleClinicalDataList;
+        // put all clinical data into sample form
+        if (CollectionUtils.isNotEmpty(patientAttributeIds)) {
+            sampleClinicalDataList = new ArrayList<>();
+            for (ClinicalData d: clinicalDataList) {
+                if (d.getSampleId() == null) {
+                    // null sample id means its a patient data, 
+                    //  we need to distribute the value to samples
+                    List<Sample> samplesForPatient = patientToSamples.get(
+                        d.getPatientId()
+                    ).get(d.getStudyId());
+                    if (samplesForPatient != null) {
+                        for (Sample s: samplesForPatient) {
+                            ClinicalData newData = new ClinicalData();
+                            newData.setAttrId(d.getAttrId());
+                            newData.setPatientId(d.getPatientId());
+                            newData.setStudyId(d.getStudyId());
+                            newData.setAttrValue(d.getAttrValue());
+                            newData.setSampleId(s.getStableId());
+                            sampleClinicalDataList.add(newData);
+                        }
+                    }
```

Comment:
If you wanted to cut down on your nesting, you could do something like this:
```suggestion
                    List<Sample> samplesForPatient = patientToSamples
                        .get(d.getPatientId())
                        .getOrDefault(d.getStudyId(), new ArrayList<>());
                    for (Sample s: samplesForPatient) {
                        ClinicalData newData = new ClinicalData();
                        newData.setAttrId(d.getAttrId());
                        newData.setPatientId(d.getPatientId());
                        newData.setStudyId(d.getStudyId());
                        newData.setAttrValue(d.getAttrValue());
                        newData.setSampleId(s.getStableId());
                        sampleClinicalDataList.add(newData);
                    }
```
This is really six in one half a dozen in the other though.